README.md
# SdarotRepo

plugin.video.sdarot.tv/changelog.txt
[B]Version 2.4.6[/B]
- Changes inside addon to fit site changes.
[B]Version 2.4.4[/B]
- Added fanart in all menus.
[B]Version 2.4.3[/B]
- requires sdarot repo.
[B]Version 2.4.2[/B]
- Changes inside addon to fit site changes(HTTPS fix by RoiD).
- Fixed: 'Attempt to fix playback crash' setting default is now OFF.
[B]Version 2.4.1[/B]
- Changes inside addon to fit site changes.
- Removed domain from addon-settings.
[B]Version 2.4.0[/B]
- Fixed: some empty categories.
- Fixed: remove unnecessary characters from summary (unescape xml characters).
[B]Version 2.3.9[/B]
- Added option to choose default max quality in addon settings (for paying members).
- Added option to right-click on item to choose quality (for paying members).
[B]Version 2.3.8[/B]
- Changes inside addon to fit site changes.
[B]Version 2.3.7[/B]
- Changes inside addon to fit site changes.
[B]Version 2.3.6[/B]
- Added waiting 30 seconds before playing an item. (changes in site).
- improve 'Clear cache' method.
- 'Clear cache' is added to addon (first row).
[B]Version 2.3.5[/B]
- 'Clear cache' is added to addon settings.
[B]Version 2.3.4[/B]
- 'Attempt to fix playback crash' setting default is now ON.
- Not requires xbmc-israel repo anymore.
[B]Version 2.3.3[/B]
- a small bugfix.
[B]Version 2.3.2[/B]
- Support cloudflare protection.
[B]Version 2.3.1[/B]
- Default domain in addon settings is now http://www.sdarot.wf
[B]Version 2.3.0[/B]
- Remove episodes from search results.
[B]Version 2.2.9[/B]
- Improvements to Kodi 15 compatibility (fixed buffering issue).
[B]Version 2.2.8[/B]
- Compatibility with Kodi 15 (off by default).
[B]Version 2.2.7[/B]
- Login to site is back.
[B]Version 2.2.6[/B]
- Quick Fix - Changes inside addon to fit site changes.
- Login to site is disabled for now (No fix for HD).
[B]Version 2.2.5[/B]
- Changes inside addon to fit site changes.
[B]Version 2.2.4[/B]
- Added program description.
[B]Version 2.2.3[/B]
- 10 tries (instead of 5) to play an item (5 seconds between each try).
[B]Version 2.2.2[/B]
- you can watch now without login.
- if server is busy, there will be 5 tries to play an item (5 seconds between each try).
[B]Version 2.2.1[/B]
-verify login before playing video (support playing video from outside the addon)
[B]Version 2.2.0[/B]
-open search screen without text in input field
-default domain in addon settings is now http://www.sdarot.pm
[B]Version 2.1.9[/B]
-watch hd if possible
[B]Version 2.1.8[/B]
-categories
[B]Version 2.1.7[/B]
-added domain to addon-settings
[B]Version 2.1.4[/B]
-minor fix
[B]Version 2.1.2[/B]
-Minor fix
[B]Version 2.1.1[/B]
-Remote domain managment
[B]Version 2.1.0[/B]
-doamin changed
-huge Timeout for very slow site.
[B]Version 2.0.9[/B]
-Site moved to new domain. increased "url timeout"
[B]Version 2.0.8[/B]
- Fix broken request, added support for Gzip for improved performance.
[B]Version 2.0.7[/B]
- Added search support - by O2ri
[B]Version 2.0.6[/B]
- 1. Modifications due to changes on the upstream website
- 2. Added new login dialog for registered users, might give better priority.
[B]Version 2.0.5[/B]
- Minor changes to script layout
[B]Version 2.0.3[/B]
- Media is now served as mp4. Changed accordingly
- Fixed chapter names appearing with "-0" appended 
[B]Version 2.0.2[/B]
- Fixed a bug - shows the chapter list of the first season as seasons instead of listing the actual seasons
[B]Version 2.0.1[/B]
- Modifications due to changes on the upstream website
[B]Version 2.0.0[/B]
- Modifications due to major rewrite on the upstream website
[B]Version 1.0.4[/B]
- Changed referer and start parameters due to some changes on the website
[B]Version 1.0.3[/B]
- Back to http
[B]Version 1.0.2[/B]
- Changed to https
[B]Version 1.0.0[/B]
- Initial release

plugin.video.sdarot.tv/service.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'resources', 'lib'))
import xbmc, xbmcaddon
from proxy import HTTP, ProxyService, PROXY_PORT

__settings__ = xbmcaddon.Addon(id='plugin.video.sdarot.tv')

class ProxyMonitor(xbmc.Monitor):
    def __init__(self):
        xbmc.Monitor.__init__(self)
        self.proxy = None
        
    def start_proxy(self):
        if self.proxy is None:
            self.proxy = ProxyService(HTTP("127.0.0.1", PROXY_PORT))
            self.proxy.start()
        
    def close_proxy(self):
        if self.proxy:
            self.proxy.stop()
            self.proxy = None

    def onSettingsChanged(self):
        use_proxy = __settings__.getSetting("use_proxy") == "true"
        if use_proxy:
            self.start_proxy()
        else:
            self.close_proxy()
            
if __name__ == '__main__':
    monitor = ProxyMonitor()
    monitor.onSettingsChanged()
    monitor.waitForAbort()  
    print 'Sdarot: closing service'
    monitor.close_proxy()
    


plugin.video.sdarot.tv/sdarottv.py
# -*- coding: utf-8 -*-

"""
	Plugin for streaming video content from www.sdarot.co.in
"""
import urllib, urllib2, re, os, sys 
import xbmcaddon, xbmc, xbmcplugin, xbmcgui
import HTMLParser
import json
import cookielib
import unicodedata

##General vars		
__plugin__ = "Sdarot.TV Video"
__author__ = "Cubicle"
__image_path__ = ''
__settings__ = xbmcaddon.Addon(id='plugin.video.sdarot.tv')
__language__ = __settings__.getLocalizedString
addonName = __settings__.getAddonInfo("name")
addonIcon = __settings__.getAddonInfo('icon')
__PLUGIN_PATH__ = __settings__.getAddonInfo('path')
LIB_PATH = xbmc.translatePath( os.path.join( __PLUGIN_PATH__, 'resources', 'lib' ) )
sys.path.append (LIB_PATH)
fanart = xbmc.translatePath(os.path.join( __PLUGIN_PATH__,"fanart.jpg"))


dbg = False # used for simple downloader logging

DOMAIN='https://www.sdarot.live'
#DOMAIN = __settings__.getSetting("domain")

#print "Sdarot Domain=" + DOMAIN

from sdarotcommon import *

path = xbmc.translatePath(__settings__.getAddonInfo("profile"))
cookie_path = os.path.join(path, 'sdarot-cookiejar.txt')
#print("Loading cookies from :" + repr(cookie_path))
cookiejar = cookielib.LWPCookieJar(cookie_path)

if os.path.exists(cookie_path):
	try:
		cookiejar.load()
	except:
		pass
elif not os.path.exists(path):
	os.makedirs(path) 
	
cookie_handler = urllib2.HTTPCookieProcessor(cookiejar)
opener = urllib2.build_opener(cookie_handler)
urllib2.install_opener(opener)
#print "built opener:" + str(opener)


def MAIN_MENU():
	CHECK_LOGIN()
	addDir('[COLOR blue]Clean chache - ניקוי מטמון[/COLOR]',"clean",7,fanart, isFolder=False)
	addDir('[COLOR red]Search - חפש[/COLOR]',DOMAIN+"/search",6,fanart)
	addDir("הכל א-ת","all-heb",2,fanart,'no');
	addDir("הכל a-z","all-eng",2,fanart,'no');
	page = getData(DOMAIN+'/series',referer=DOMAIN)
	matches = re.compile('text="בחירת ז\'אנר">(.*?)</select>', re.S).findall(page)
	matches = re.compile('<option value="(.*?)" >(.*?)</option>').findall(matches[0])
	for match in matches:
		 addDir(str(match[1]),"all-heb",2,fanart,str(match[0]))
	
def SearchSdarot(url,search_entered):
	search_entered= ''
	keyboard = xbmc.Keyboard("", "חפש כאן")
	keyboard.doModal()
	if keyboard.isConfirmed():
		search_entered = keyboard.getText()
	page = getData(url=url,timeout=0,postData="term=" + search_entered)
	
	matches = re.compile('href="/watch/(\d+)-(.*?)">').findall(page)

	#needs to remove duplicted result (originaly in site
	matches = set(matches)
	#print matches
	for match in matches:
		series_id = match[0]
		link_name = match[1]
		image_link = DOMAIN+"/media/series/"+str(match[0])+".jpg"
		series_link = DOMAIN+"/watch/"+str(match[0])+"/"+match[1]
		if 'episode' not in link_name:
			addDir(link_name,series_link,"3&image="+urllib.quote(image_link)+"&series_id="+series_id+"&series_name="+urllib.quote(link_name),image_link)
		
def INDEX_AZ(url,page):
	if page == 'no':
		page = 'https://www.sdarot.pm/ajax/series?loadMore=9999&start=1&search%5Bfrom%5D=&search%5Bto%5D=&search%5Border%5D=&search%5Bdir%5D='
	else:
		page = 'https://www.sdarot.pm/ajax/series?loadMore=9999&start=1&search%5Bgenre%5D%5B%5D={0}&search%5Bfrom%5D=&search%5Bto%5D=&search%5Border%5D=&search%5Bdir%5D='.format(page)
	page = getData(page)
	matches = re.compile('<img.*?src="(.*?)".*?<h4>(.*?)</h4>.*?<h5>(.*?)</h5>.*?href="(.*?)"', re.S).findall(page)
	sr_arr = []
	for match in matches:
		heb_name = HTMLParser.HTMLParser().unescape(match[1].decode("utf-8")).strip()
		eng_name = HTMLParser.HTMLParser().unescape(match[2].decode("utf-8")).strip()
		if url == "all-eng":
			sr_arr.append(( match[0], match[3], eng_name ))
		else:
			sr_arr.append(( match[0], match[3], heb_name ))
	sr_arr.sort(key=lambda x: x[2])

	for key in sr_arr:
		series_link = DOMAIN + str(key[1])
		image_link = 'https:' + str(key[0])
		series_id = image_link[image_link.rfind('/')+1:image_link.rfind('.')]
		addDir(key[2],series_link,"3&image="+urllib.quote(image_link)+"&series_id="+series_id+"&series_name="+urllib.quote(key[2].encode("utf-8")),image_link)
	xbmcplugin.setContent(int(sys.argv[1]), 'tvshows')
	  
def sdarot_series(url):
	series_id=urllib.unquote_plus(params["series_id"])
	series_name=urllib.unquote_plus(params["series_name"])
	image_link=urllib.unquote_plus(params["image"])
	try:
		page = getData(url,referer=DOMAIN+'/series')
	except urllib2.URLError, e:
		print 'sdarot_season: got http error ' +str(e.code) + ' fetching ' + url + "\n"
		raise e
	matches = re.compile('תקציר הסדרה.*?<p>(.*?)</p>', re.S).findall(page)
	summary = matches[0] if len(matches) == 1 else ""

	seasons_list = re.compile('id="season">(.*?)</ul>',re.S).findall(page)[0]
	matches = re.compile('>(\d+)</a').findall(seasons_list)
			
	for season in matches:
		addDir("עונה "+ str(season),url,"5&image="+urllib.quote(image_link)+"&season_id="+str(season)+"&series_id="+str(series_id)+"&series_name="+urllib.quote(series_name),image_link,summary=summary)
	#xbmcplugin.setContent(int(sys.argv[1]), 'tvshows')
	xbmcplugin.setContent(int(sys.argv[1]), 'episodes')
	xbmc.executebuiltin('Container.SetViewMode(504)')
	  
def sdarot_season(url, summary):
	series_id=urllib.unquote_plus(params["series_id"])
	series_name=urllib.unquote_plus(params["series_name"])
	season_id=urllib.unquote_plus(params["season_id"])
	image_link=urllib.unquote_plus(params["image"])
	page = getData(url=DOMAIN+"/ajax/watch?episodeList="+series_id+"&season="+season_id,timeout=0,referer=url);
	episodes = re.compile('<li data-episode="(.*?)"').findall(page)
	if not episodes:
		xbmcgui.Dialog().ok('Error occurred',"לא נמצאו פרקים לעונה")
		return
	
	for epis in episodes:
		addVideoLink("פרק "+epis, url, "4&episode_id="+epis+"&image="+urllib.quote(image_link)+"&season_id="+str(season_id)+"&series_id="+str(series_id)+"&series_name="+urllib.quote(series_name),image_link, summary)		 
	xbmcplugin.setContent(int(sys.argv[1]), 'episodes')
	xbmc.executebuiltin('Container.SetViewMode(504)')
		
def sdarot_movie(url, summary, m_quality=None):
	referer=url
	series_id=urllib.unquote_plus(params["series_id"])
	series_name=urllib.unquote_plus(params["series_name"])
	season_id=urllib.unquote_plus(params["season_id"])
	image_link=urllib.unquote_plus(params["image"])
	episode_id=urllib.unquote_plus(params["episode_id"])
	title = series_name + ", עונה " + season_id + ", פרק " + episode_id
	i = url.rfind('/')
	url = '{0}-{1}/season/{2}/episode/{3}'.format(url[:i], urllib.quote(url[i+1:]), season_id, episode_id)
	finalUrl, VID = getFinalVideoUrl(series_id, season_id, episode_id, url, m_quality=m_quality)
	page = getData(url=DOMAIN+"/ajax/watch",timeout=1,postData="count="+VID,referer=referer)
	liz = xbmcgui.ListItem(path=finalUrl)
	liz.setInfo(type="Video", infoLabels={ "Title": title, "Plot": urllib.unquote(summary) })	
	liz.setProperty('IsPlayable', 'true')
	xbmcplugin.setResolvedUrl(handle=int(sys.argv[1]), succeeded=True, listitem=liz)
	ok = xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=finalUrl, listitem=liz, isFolder=False)

def choose_quality(url, summary):
	series_id=urllib.unquote_plus(params["series_id"])
	series_name=urllib.unquote_plus(params["series_name"])
	season_id=urllib.unquote_plus(params["season_id"])
	image_link=urllib.unquote_plus(params["image"])
	episode_id=urllib.unquote_plus(params["episode_id"])
	title = series_name + ", עונה " + season_id + ", פרק " + episode_id
	qualities = getFinalVideoUrl(series_id, season_id, episode_id, url, m_quality='choose')
	for q in qualities:
		addVideoLink(title+" ("+ str(q) +")",url,"4&episode_id="+episode_id+"&image="+urllib.quote(image_link)+"&season_id="+str(season_id)+"&series_id="+str(series_id)+"&series_name="+urllib.quote(series_name)+"&quality="+str(q),image_link,summary,contextMenu=False)
	xbmcplugin.setContent(int(sys.argv[1]), 'episodes')
	xbmc.executebuiltin('Container.SetViewMode(504)')

def clean_cache():
	try:
		if os.path.isfile(cookie_path):
			os.unlink(cookie_path)
	except Exception as ex:
		xbmc.log(str(ex), 3)
	tempDir = xbmc.translatePath('special://temp/').decode("utf-8")
	for the_file in os.listdir(tempDir):
		if not '.fi' in the_file and the_file != 'cookies.dat':
			continue
		file_path = os.path.join(tempDir, the_file)
		try:
			if os.path.isfile(file_path):
				os.unlink(file_path)
		except Exception as ex:
			xbmc.log(str(ex), 3)
	
params = getParams(sys.argv[2])
#print "params:"
#print params
url=None
name=None
mode=None
module=None
page=None
summary=None
quality=None

try:
	url=urllib.unquote_plus(params["url"])
except:
	pass
try:
	name=urllib.unquote_plus(params["name"])
except:
	pass
try:
	mode=int(params["mode"])
except:
	pass
try:
	module=urllib.unquote_plus(params["module"])
except:
	pass
try:
	page=urllib.unquote_plus(params["page"])
except:
	pass
try:
	summary=urllib.unquote_plus(params["summary"])
except:
	pass
try:
	quality=int(params["quality"])
except:
	pass

if mode==None or url==None or len(url)<1:
	MAIN_MENU()

elif mode==2:
	INDEX_AZ(url,module)
elif mode==3:
	sdarot_series(url)
elif mode==4:
	sdarot_movie(url, summary, quality)
elif mode==5:
	sdarot_season(url, summary)
elif mode==6:
	SearchSdarot(url,name)
elif mode==7:
	clean_cache()
	xbmc.executebuiltin("XBMC.Notification({0}, המטמון נמחק, {1}, {2})".format(addonName, 5000 ,addonIcon))
elif mode==8:
	choose_quality(url, summary)

xbmcplugin.setPluginFanart(int(sys.argv[1]),xbmc.translatePath( os.path.join( __PLUGIN_PATH__,"fanart.jpg") ))
xbmcplugin.endOfDirectory(int(sys.argv[1]))


plugin.video.sdarot.tv/resources/lib/sdarotcommon.py
# -*- coding: utf-8 -*-

'''
Created on 30/04/2011

@author: shai
'''
__USERAGENT__ = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36'
#__USERAGENT__ = 'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.1; WOW64; Trident/6.0)'
DOMAIN='https://www.sdarot.live'


import urllib,urllib2,re,xbmc,xbmcplugin,xbmcgui,xbmcaddon,os,sys,time, socket
import StringIO
import gzip
import json
from operator import itemgetter, attrgetter
from proxy import PROXY_PORT
import cloudflare

__settings__ = xbmcaddon.Addon(id='plugin.video.sdarot.tv')
__language__ = __settings__.getLocalizedString
__cachePeriod__ = __settings__.getSetting("cache")
__PLUGIN_PATH__ = __settings__.getAddonInfo('path')
__DEBUG__ = __settings__.getSetting("DEBUG") == "true"
#__DEBUG__ = False

#DOMAIN = __settings__.getSetting("domain")
HOST = DOMAIN[8:]
#PROXY_PORT = 9899

#print "common domain=" +  DOMAIN
#print "common domain="+ HOST

__REFERER__ = DOMAIN+'/templates/frontend/blue_html5/player/jwplayer.flash.swf'

def LOGIN():
	#print("LOGIN  is running now!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
	loginurl = DOMAIN+'/login'
	if __settings__.getSetting('username')=='':
		dialog = xbmcgui.Dialog()
		xbmcgui.Dialog().ok('Sdarot','www.sdarot.tv התוסף דורש חשבון  חינמי באתר' ,' במסך הבא יש להכניס את שם המשתמש והסיסמא')
		search_entered = ''
		keyboard = xbmc.Keyboard(search_entered, 'נא הקלד שם משתמש')
		keyboard.doModal()
		if keyboard.isConfirmed():
			search_entered = keyboard.getText() 
		__settings__.setSetting('username',search_entered)
		
	if __settings__.getSetting('user_password')=='':
		search_entered = ''
		keyboard = xbmc.Keyboard(search_entered, 'נא הקלד סיסמא')
		keyboard.doModal()
		if keyboard.isConfirmed():
			search_entered = keyboard.getText()
		__settings__.setSetting('user_password',search_entered)

	username = __settings__.getSetting('username')
	password = __settings__.getSetting('user_password')
	if not username or not password:
		print "Sdarot tv:no credencials found skipping login"
		return
	
	print "Trying to login to sdarot tv site username:" + username
	page = getData(url=loginurl,timeout=0,postData="username=" + username + "&password=" + password +"&submit_login=התחבר",referer=DOMAIN);
   
def CHECK_LOGIN():
	# check's if login  is required.
	#print "check if logged in already"
	if __settings__.getSetting('username').strip() == '' or __settings__.getSetting('user_password') == '':
		return
		
	page = getData(DOMAIN+'/series',referer=DOMAIN)
	#print page
	match = re.compile('<span class="button blue" id="logout"><a href=".*?/log(.*?)">').findall(page)
	#print match
	if match:
		if str(match[0])!='out':
		#	print "login required"
			LOGIN()
		#else:
		#	print "already logged in."
	else:
		LOGIN()
 
def enum(**enums):
		return type('Enum', (), enums)

def getMatches(url, pattern):
		page = getData(url)
		matches=re.compile(pattern).findall(page)
		return matches   

def getParams(arg):
		param=[]
		paramstring=arg
		if len(paramstring)>=2:
			params=arg
			cleanedparams=params.replace('?','')
			pairsofparams=cleanedparams.split('&')
			param={}
			for elem in pairsofparams:
				splitparams=elem.split('=')
				if (len(splitparams))==2:	
					param[splitparams[0]]=splitparams[1]
		return param
	
def addDir(name, url, mode, iconimage='DefaultFolder.png', elementId=None, summary='', fanart='',contextMenu=None, isFolder=True):
	if fanart=='':
		fanart = xbmc.translatePath(os.path.join( __PLUGIN_PATH__,"fanart.jpg"))
	u = sys.argv[0] + "?url=" + urllib.quote_plus(url) + "&mode=" + str(mode) + "&name=" + name+ "&summary=" + urllib.quote_plus(summary)
	if elementId:
		u += "&module=" + urllib.quote_plus(elementId)
	liz = xbmcgui.ListItem(name, iconImage=iconimage, thumbnailImage=iconimage)
	liz.setInfo(type="Video", infoLabels={ "Title": urllib.unquote(name), "Plot": UnEscapeXML(urllib.unquote(summary))})
		
	if contextMenu:
		liz.addContextMenuItems(items=contextMenu, replaceItems=False)

	if fanart:
		liz.setProperty("Fanart_Image", fanart)
	ok = xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=u, listitem=liz, isFolder=isFolder)
	return ok

def addVideoLink(name, url, mode, iconimage='DefaultFolder.png', summary = '', fanart='', contextMenu=True):
	if fanart=='':
		fanart = xbmc.translatePath(os.path.join( __PLUGIN_PATH__,"fanart.jpg"))
	u = sys.argv[0] + "?url=" + urllib.quote_plus(url) + "&mode=" + str(mode) + "&name=" + "&summary=" + urllib.quote_plus(summary)
	liz = xbmcgui.ListItem(name, iconImage=iconimage, thumbnailImage=iconimage)
	liz.setInfo(type="Video", infoLabels={ "Title": urllib.unquote(name), "Plot": UnEscapeXML(urllib.unquote(summary))})	
	liz.setProperty("Fanart_Image", fanart)
	liz.setProperty('IsPlayable', 'true')
	if contextMenu:
		liz.addContextMenuItems(items=[(__language__(30005).encode('utf-8'), 'XBMC.Container.Update({0})'.format(u.replace('mode=4', 'mode=8')))])
	ok = xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=u, listitem=liz, isFolder=False)
	return ok
	
def addLink(name, url, iconimage='DefaultFolder.png', sub=''):
	liz = xbmcgui.ListItem(name, iconImage="DefaultVideo.png", thumbnailImage=iconimage)
	liz.setInfo(type="Video", infoLabels={ "Title": urllib.unquote(name), "Plot": urllib.unquote(sub)})
	ok = xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=liz)
	return ok

def extractFromZip(gzipData):
	data = StringIO.StringIO(gzipData)
	gzipper = gzip.GzipFile(fileobj=data)
	html = gzipper.read()
	gzipper.close()
	return html
	
def getData_attempt(url, timeout=__cachePeriod__, name='', postData=None,referer=__REFERER__):
		#print 'getData: url --> ' + url + '\npostData-->' + str(postData)
		if __DEBUG__:
			print 'name --> ' + name
		#temporary disabled the cache - cause problems with headers	
		if timeout > 9999999:
			if name == '':
				cachePath = xbmc.translatePath(os.path.join(__PLUGIN_PATH__, 'cache', 'pages', urllib.quote(url,"")))
			else:
				cachePath = xbmc.translatePath(os.path.join(__PLUGIN_PATH__, 'cache', 'pages', name))
			if (os.path.exists(cachePath) and (time.time()-os.path.getmtime(cachePath))/60/60 <= float(timeout)):
				with open(cachePath) as f:
					ret = f.read()
					if __DEBUG__:
						print 'returned data from cache'
					return ret
		socket.setdefaulttimeout(15)
		req = urllib2.Request(url)
		req.add_header('User-Agent', __USERAGENT__)   
		req.add_header('X-Requested-With','XMLHttpRequest')
		req.add_header('Accept','application/json, text/javascript, */*; q=0.01')
		req.add_header('Accept-Encoding','gzip,deflate,sdch')
		req.add_header('Content-Type','application/x-www-form-urlencoded; charset=UTF-8')
		req.add_header('Connection','keep-alive')
		req.add_header('Host',HOST)
		req.add_header('Origin',DOMAIN)
		if postData:
			req.add_header('Content-Length',len(postData))
		
		if referer: 
			req.add_header ('Referer',referer)
			
		if __DEBUG__:
			print "sent headers:" + str(req.headers)           
#		response = urllib2.urlopen(url=req,timeout=180,data=postData)
		response = cloudflare.ddos_open(url=req,timeout=180,data=postData)

		if __DEBUG__:
			print "received headers:" + str(response.info());
		
		if response.info().get('Content-Encoding') == 'gzip':
			buf = StringIO.StringIO( response.read())
			f = gzip.GzipFile(fileobj=buf)
			data = f.read()
			#print "received gzip len " + str(len(data))
		else:
			data = response.read()

		if data:
			data = data.replace("\n","").replace("\t","").replace("\r","")   
					 
		try:
			#print sys.modules["__main__"].cookiejar
			sys.modules["__main__"].cookiejar.save()
			
		except Exception,e:
			print e	   
		
		if __DEBUG__:
			print "recieved data:" + str(data)
					   
		response.close()
		
		try:
			if timeout > 999999:
				with open(cachePath, 'wb') as f:
					f.write(data)
			if __DEBUG__:
				print data
			return data
		except:
			return data
	
def getData(url, timeout=__cachePeriod__, name='', postData=None,referer=__REFERER__):
		for i in range(3):
		  #print "getData: Attempt " + str(i)
		  try:
			return getData_attempt(url, timeout, name, postData,referer)
		  except urllib2.URLError, e:
			print e
			if (i == 2):
			  raise e

def getFinalVideoUrl(series_id,season_id,episode_id,referer,silent=False,m_quality=None):
	CHECK_LOGIN()
	if m_quality and m_quality != 'choose':
		max_quality = m_quality
	else:
		max_quality = int(__settings__.getSetting("max_quality"))
	for i in range(10):
		error = ''
		tok = getData(url=DOMAIN+"/ajax/watch",timeout=1,postData="preWatch=true&SID="+series_id+"&season="+season_id+"&ep="+episode_id,referer=referer)
		try:
			page = getData(url=DOMAIN+"/ajax/watch",timeout=1,postData="watch=true&token="+tok+"&serie="+series_id+"&season="+season_id+"&episode="+episode_id,referer=referer)
			prms=json.loads(page)
			if "error" in prms:
				dp = xbmcgui.DialogProgress()
				dp.create("לצפייה באיכות HD וללא המתנה ניתן לרכוש מנוי", "אנא המתן 30 שניות", '', "[COLOR orange][B]לרכישת מנוי להיכנס בדפדפן - www.sdarot.click/donate[/B][/COLOR]")
				dp.update(0)
				for s in range(30,-1,-1):
					time.sleep(1)
					dp.update(int((30-s)/30.0 * 100), "אנא המתן 30 שניות", 'עוד {0} שניות'.format(s), '')
					if dp.iscanceled(): 
						dp.close()
						return None,None
				page = getData(url=DOMAIN+"/ajax/watch",timeout=1,postData="watch=true&token="+tok+"&serie="+series_id+"&season="+season_id+"&episode="+episode_id,referer=referer)
		except Exception as e:
			pass
		
		token = None
		
		try:
			prms=json.loads(page)
			if "error" in prms:
				error = str(prms["error"].encode("utf-8"))
				if len(error) > 0 :
					time.sleep(5)
					continue
			
			vid_url = str(prms["url"])
			
			quality = 0
			watch = prms["watch"]
			qualities = []
			for k, v in watch.iteritems():
				qualities.append(k)
				k = int(k)
				if quality < k <= max_quality:
					quality = k
					token = v
			if m_quality == 'choose':
				return qualities
			
			VID = str(prms["VID"])
			vid_time = str(prms["time"])
			break
		
		except Exception as e:
			error = e
			time.sleep(5)

	if error:
		xbmc.log('error:'+str(error), 3)
		if not silent:
			xbmcgui.Dialog().ok('Error occurred',str(error))
		return None,None
		
	if not token:
		if not silent:
			xbmcgui.Dialog().ok('Error occurred',"התוסף לא הצליח לקבל אישור לצפייה, אנא נסה מאוחר יותר")
		return None,None
	finalUrl = "http://" + vid_url + "/watch/" + str(quality) + "/" + VID + '.mp4?token=' + token + '&time=' + vid_time
	
	if __settings__.getSetting("use_proxy") == "true":
		finalUrl = "http://127.0.0.1:{0}/?url={1}".format(PROXY_PORT, urllib.quote(finalUrl))

	return "{0}|Referer={1}&User-Agent={2}".format(finalUrl, urllib.quote(referer), __USERAGENT__), VID
	#return finalUrl, VID

def getImage(imageURL, siteName):
		imageName = getImageName(imageURL)
		cacheDir = xbmc.translatePath(os.path.join(__PLUGIN_PATH__, 'cache', 'images', siteName))
		cachePath = xbmc.translatePath(os.path.join(cacheDir, imageName))
		if not os.path.exists(cachePath):
			## fetch the image and store it in the cache path
			if not os.path.exists(cacheDir):
				os.makedirs(cacheDir)
			urllib.urlretrieve(imageURL, cachePath)
		return cachePath
		
def getImageName(imageURL):
		idx = int(imageURL.rfind("/")) + 1
		return imageURL[idx:]

def UnEscapeXML(str):
	return str.replace('&amp;', '&').replace("&lt;", "<").replace("&gt;", ">").replace("&quot;", '"').replace("&#39;", "'").replace("&#039;", "'")
